Everest mount write docs (#8725)
treeverse/lakeFS: docs/reference/mount-write-mode-semantics.md

Reviewer: Im missing a section that explains how sync is done, with specific details: 
- How upload works 
- Caching writes and invalidation
- No eviction

Author: I really not sure what you mean by that
How upload works? In what context?
How do these internal details affect the user?

## lakeFS Consistency
Local changes are reflected in lakeFS only after the changes are **committed**.
That means that the data is not available for other users until the changes are committed.
If, for example, two users mount the same branch, they will not see each other's changes until they are committed.

# Mount Write Mode File System Behavior

## Functionality Limitations
- Newly created empty directories will not reflect as directory marker in lakeFS.
- Mount write mode is not supported with Everest partial-read mode.

## File System Behavior
- Rename is not supported.
- Temporary files are not supported.
- Hard/symbolic links are not supported.
- POSIX file locks (lockf) are not supported.
- POSIX permissions are not supported- default permissions are assigned given to files and dirs.
- A file system directory cannot contain both a file and a directory of the same name (e.g., dir and dir/file).
- A deleted file's name cannot be used as a directory type later and the same for opposite types (e.g, Not allowed: touch foo; rm foo; mkdir foo;).
- Calling remove on a directory type will fail explicitly with an error.

## Behavior modified:

> - File opened for Read And Write will be treated as Opened for Write.

Reviewer: remove, irrelevant anymore

> - Modifying file metadata (chmod, chown, chgrp, time) will result in an error unless the file doesn't exist.

Reviewer: noop unless file doesnt exist - then error

---
title: Mount Write Mode Semantics
description: Everest mount file system semantics, limitations and and general info.
grand_parent: Features
parent: Mount
---
# Mount Write Mode Semantics

{: .d-inline-block }
experimental
{: .label .label-red }

## Consistency Model

### File System Consistency

> Everest mount provides a strong read-after-write consistency model. 

Reviewer 2: ```suggestion
Everest mount provides a strong read-after-write consistency model within a single mount point. 
```

just to clarify that this doesn't happen if we mount the same path at different locations

This means that once a write operation is done, the data is guaranteed to be available for subsequent read operations.

### lakeFS Consistency
Local changes are reflected in lakeFS only after the changes are **committed**.
That means that the data is not available for other users until the changes are committed.
If, for example, two users mount the same branch, they will not see each other's changes until they are committed.

## Mount Write Mode File System Behavior

### Functionality Limitations
- Newly created empty directories will not reflect as directory marker in lakeFS.

> - LakeFS allows having 2 path keys that one is a "directory" prefix of the other, for example the following 2 lakeFS keys are valid: `animals/cat.png` and `animals` (empty object) but since a file system cannot contain both a file and a directory of the same name it will lead to an undefined behavior depending on the Filesystem type (e.g., dir and dir/file).

Reviewer 3: LakeFS?
Shame Shame Shame
```suggestion
- lakeFS allows having 2 path keys that one is a "directory" prefix of the other, for example the following 2 lakeFS keys are valid: `animals/cat.png` and `animals` (empty object) but since a file system cannot contain both a file and a directory of the same name it will lead to an undefined behavior depending on the Filesystem type (e.g., dir and dir/file).
```

### File System Behavior

#### Not Supported:
- Rename is not supported.
- Temporary files are not supported.
- Hard/symbolic links are not supported.
- POSIX file locks (lockf) are not supported.

> - POSIX permissions are not supported- default permissions are given to files and dirs.

Reviewer 2: ```suggestion
- POSIX permissions are not supported - default permissions are given to files and dirs.
```

- A deleted file's name cannot be used as a directory type later and the same for opposite types (e.g, Not allowed: touch foo; rm foo; mkdir foo;).
- Calling remove on a directory type will fail explicitly with an error.

#### Behavior modified:
- Modifying file metadata (chmod, chown, chgrp, time) will result in noop (the file metadata will not be changed). 
- Remove:

>   - (a) Open handlers to a file will error if the file is removed during a read operation.

Reviewer: Delete this line, incorrect

>   - (b) Removal is not an atomic operation, calling remove and open at the same time might result in a race condition where the open might succeed.

Reviewer: ```suggestion
  - When calling remove we mark a file as a tombstone using  [Extended Attributes](https://en.wikipedia.org/wiki/Extended_file_attributes) API's.
  - Removal is not an atomic operation, calling remove and open at the same time might result in a race condition where the open might succeed.
```
